src/functions.php

```diff
@@ -25,15 +25,29 @@ function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
             $loop->cancelTimer($timer);
             $reject($v);
         });
+    }, function ($resolveUnused, $reject) use ($loop, &$timer, $promise) {
+        // cancelling this promise will cancel the timer and reject
+        $loop->cancelTimer($timer);
+        $reject(new \RuntimeException('Timeout cancelled'));
```

Comment:
This is an interesting idea – albeit different to what we've discussed in #3 so far :)

I'll take a look at this and see how this works out.

This has some different implications than the current implementation. In particular, we can no longer rely on the outcome of the `cancel()` call (though this may not necessarily be a bad thing as you've pointed out).